kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/utils/ByteBufUtils.java

```diff
@@ -117,9 +117,10 @@ public static DecodeResult decodePulsarEntryToKafkaRecords(final MessageMetadata
                     baseOffset,
                     metadata.getPublishTime(),
                     0,
-                    metadata.getTxnidMostBits(),
-                    (short) metadata.getTxnidLeastBits(),
-                    new EndTransactionMarker(controlRecordType, 0)
+                    metadata.hasTxnidMostBits() ? metadata.getTxnidMostBits() : Long.MAX_VALUE,
+                    metadata.hasTxnidLeastBits() ? (short) metadata.getTxnidLeastBits() : 0,
```

Comment:
Just a concern for the original logic. Why pass the `txnidMostBits` as the producer id and `txnidLeastBits` as the epoch?

```java
    public static MemoryRecords withEndTransactionMarker(
            long initialOffset, long timestamp, int partitionLeaderEpoch,
            long producerId, short producerEpoch, EndTransactionMarker marker) {
```

Reply:
You can see the `PulsarEntryFormatter`, we will convert the records as pulsar messages, and the Kafka transaction marker will be treated as a pulsar transaction marker.

```
records.batches().forEach(recordBatch -> {
    boolean controlBatch = recordBatch.isControlBatch();
    StreamSupport.stream(recordBatch.spliterator(), true).forEachOrdered(record -> {
        MessageImpl<byte[]> message = recordToEntry(record);
        messages.add(message);
        if (recordBatch.isTransactional()) {
            msgMetadata.setTxnidMostBits(recordBatch.producerId());
            msgMetadata.setTxnidLeastBits(recordBatch.producerEpoch());
        }
        if (controlBatch) {
            ControlRecordType controlRecordType = ControlRecordType.parse(record.key());
            switch (controlRecordType) {
                case ABORT:
                    msgMetadata.setMarkerType(MarkerType.TXN_ABORT_VALUE);
                    break;
                case COMMIT:
                    msgMetadata.setMarkerType(MarkerType.TXN_COMMIT_VALUE);
                    break;
                default:
                    msgMetadata.setMarkerType(MarkerType.UNKNOWN_MARKER_VALUE);
                    break;
            }
        }
    });
});
```